File: 139_Word Break.cpp
/*
139. Word Break
Given a string s and a dictionary of words, determine if s can be segmented into a space-separated sequence of one or more dictionary words.
Use DP where dp[i] is true if s[:i] can be segmented.
*/
#include <bits/stdc++.h>
using namespace std;

bool wordBreak(string s, vector<string>& wordDict) {
    unordered_set<string> dict(wordDict.begin(), wordDict.end());
    int n = s.size();
    vector<bool> dp(n+1,false);
    dp[0] = true;
    for (int i = 1; i <= n; ++i) {
        for (int j = 0; j < i; ++j) {
            if (dp[j] && dict.count(s.substr(j, i-j))) { dp[i] = true; break; }
        }
    }
    return dp[n];
}

int main() {
    string s = "leetcode";
    vector<string> dict{"leet","code"};
    cout << "wordBreak('leetcode') = " << (wordBreak(s, dict) ? "true" : "false") << "\n"; // true
    string t = "applepenapple";
    vector<string> dict2{"apple","pen"};
    cout << "wordBreak('applepenapple') = " << (wordBreak(t, dict2) ? "true" : "false") << "\n"; // true
    return 0;
}

// Time Complexity: O(n^2 * k) substring cost, can be improved
// Space Complexity: O(n)

File: 139_Word Break.py
"""
139. Word Break
Dynamic programming to check if s can be segmented using words in dict.
"""
from typing import List

def word_break(s: str, wordDict: List[str]) -> bool:
    dict_set = set(wordDict)
    n = len(s)
    dp = [False] * (n+1)
    dp[0] = True
    for i in range(1, n+1):
        for j in range(i):
            if dp[j] and s[j:i] in dict_set:
                dp[i] = True
                break
    return dp[n]

if __name__ == '__main__':
    print("wordBreak('leetcode') =", word_break('leetcode', ['leet','code']))
    print("wordBreak('applepenapple') =", word_break('applepenapple', ['apple','pen']))

# Time Complexity: O(n^2 * k) where k is substring check cost
# Space Complexity: O(n)
